src/commands/search/search-packages-hinthack.cc

```diff
@@ -6,50 +6,191 @@
 
-#include "Zypper.h"
-
 #include "search-packages-hinthack.h"
 
+#include "Zypper.h"
+#include "subcommand.h"
+
 using namespace zypp;
 
-void SLE15_SearchPackagesHintHack( Zypper & zypper )
+#undef	ENABLE_COULD_BE_INSTALLED_HINT // let's see if we need it
+#undef	ENABLE_DISABLED_IN_CONFIG_HINT // let's see if we need it
+
+///////////////////////////////////////////////////////////////////
+namespace searchPackagesHintHack
 {
-  if ( ! zypper.config().do_ttyout )
-    return;
+  ///////////////////////////////////////////////////////////////////
+  namespace
+  {
+#undef	FAKED
+#ifdef  FAKED
+#warning FAKE
+#warning FAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAKEFAFAKEFAKEKEFAKE
+#warning FAKE
+    static const std::string ZSPP_PackageName { "zypper" };
+    static const Pathname    ZSPP_BinaryPath  { "/usr/lib/zypper/commands/zypper-search-packages" };
+    static const Capability  ZSPP_CliFwdCap   { "y2pmsh" };
+#else
+    static const std::string ZSPP_PackageName { "zypper-search-packages-plugin" };
+    static const Pathname    ZSPP_BinaryPath  { "/usr/lib/zypper/commands/zypper-search-packages" };
+    static const Capability  ZSPP_CliFwdCap   { "supports-zypper-cli-forwarding" };
+#endif
+    /** Whether hinting/running search-packages should be considered at all. */
+    bool maySearchPackagesAtAll( Zypper & zypper_r )
+    {
+      if ( ! zypper_r.config().do_ttyout )
+	return false;	// no terminal output, i.e. no user in front
+
+      Out & out( zypper_r.out() );
+      if ( ! out.typeNORMAL() || out.verbosity() < Out::NORMAL )
+	return false;	// xml output or quiet mode
+
+      if ( zypper_r.runningShell() )
+	return false;	// we don't support subcommands in shell
+
+      if ( zypper_r.globalOpts().non_interactive )
+	return false;	// Never forward in non-interactive mode (search-packages itself relies on this!)
+
+      if ( zypper_r.globalOpts().changedRoot )
+	return false;	// we're chrooted
+
+      return true;
+    }
+
+    /** Whether an installed search-packages subcommand provides \c supports-zypper-cli-forwarding.
+     * FATE#325599: Introduced with SLE15-SP1 and forces us to even call the plugin.
+     */
+    bool searchPackagesSupportsZypperCliForwarding()
+    {
+      using target::rpm::librpmDb;
+      // RpmDb access is blocked while the Target is not initialized.
+      // Launching the Target just for this query would be an overkill.
+      struct TmpUnblock {
+	TmpUnblock()
+	: _wasBlocked( librpmDb::isBlocked() )
+	{ if ( _wasBlocked ) librpmDb::unblockAccess(); }
+	~TmpUnblock()
+	{ if ( _wasBlocked ) librpmDb::blockAccess(); }
+      private:
+	bool _wasBlocked;
+      } tmpUnblock;
+
+      librpmDb::db_const_iterator it;
+      return( it.findPackage( ZSPP_PackageName ) && it->tag_provides().count( ZSPP_CliFwdCap ) );
+    }
+
+    /** Print a hint about an available search-packages subcommand. */
+    void printSearchPackagesHint( Zypper & zypper_r, bool plgInstalled_r )
```

Comment:
Zypper & zypper_r -> didn't you say that convention is Zypper & zypper? I don't mind either but would be good to know before I clean up the  Zypper &zypp_r  in the commands PR

Reply:
I wanted to demonstrate the use of `_r` and overdid it :)

Reply:
BTW in `Zypper &zypp_r ` I nagged about `zypp` (missing `er`), not so much about the `_r`.  The code uses both `zypper` and `zypper_r`.